ring/partition_ring.go

```diff
@@ -340,3 +352,60 @@ func (r *PartitionRing) String() string {
 
 	return fmt.Sprintf("PartitionRing{ownersCount: %d, partitionsCount: %d, partitions: {%s}}", len(r.desc.Owners), len(r.desc.Partitions), buf.String())
 }
+
+// ActivePartitionBatchRing wraps PartitionRing and implements DoBatchRing to lookup ACTIVE partitions.
+type ActivePartitionBatchRing struct {
+	ring *PartitionRing
+}
+
+func NewActivePartitionBatchRing(ring *PartitionRing) *ActivePartitionBatchRing {
+	return &ActivePartitionBatchRing{
+		ring: ring,
+	}
+}
+
+// InstancesCount returns the number of active partitions in the ring.
+//
+// InstancesCount implements DoBatchRing.InstancesCount.
+func (r *ActivePartitionBatchRing) InstancesCount() int {
+	return r.ring.ActivePartitionsCount()
+}
+
+// ReplicationFactor returns 1 as partitions replication factor: an entry (looked by key via Get())
+// is always stored in 1 and only 1 partition.
+//
+// ReplicationFactor implements DoBatchRing.ReplicationFactor.
+func (r *ActivePartitionBatchRing) ReplicationFactor() int {
+	return 1
+}
+
+// Get implements DoBatchRing.Get.
+func (r *ActivePartitionBatchRing) Get(key uint32, _ Operation, bufInstances []InstanceDesc, _, _ []string) (ReplicationSet, error) {
+	partitionID, err := r.ring.ActivePartitionForKey(key)
+	if err != nil {
+		return ReplicationSet{}, err
+	}
+
+	// Ensure we have enough capacity in bufInstances.
+	if cap(bufInstances) < 1 {
+		bufInstances = []InstanceDesc{{}}
+	} else {
+		bufInstances = bufInstances[:1]
+	}
+
+	partitionIDString := strconv.Itoa(int(partitionID))
+
+	bufInstances[0] = InstanceDesc{
+		Addr:      partitionIDString,
+		Timestamp: 0,
+		State:     ACTIVE,
+		Id:        partitionIDString,
+	}
+
+	return ReplicationSet{
+		Instances:            bufInstances,
```

Comment:
Since we return instances in shared buffer, `ReplicaSet` is only valid until next call to `Get` with the same buffer. This is very tricky behaviour, and should be documented. (`DoBatch` is doing the right thing -- it uses `ReplicaSet.Instances` and then doesn't use same `ReplicaSet`.)

If we moved buffer into `ActivePartitionBatchRing` as suggested above, this would still be tricky, but at least it would be clearer which buffer is "the same".

Reply:
I think we should document that buffer is used, but for the rest of the comment I see it differently:

- Given buffer is explicitly passed from outside, it's in caller control whether to reuse it or not (but I agree should be documented)
- The adapter approach, instead, mask it and it's outside of caller control whether a buffer should be reused or not

Reply:
Explained buffers in the interface doc: https://github.com/grafana/dskit/pull/484/commits/1f95b9219b6c4e59da02c45c7420233b65c85595